storage: TestReplicaCancelRaft is disabled

SHA: https://github.com/cockroachdb/cockroach/commits/fc842bc609b6a937998e5366703d7d318544169c

Stress build found a failed test:

```
I161106 07:42:05.612538 5158 gossip/gossip.go:237  [n?] initial resolvers: []
W161106 07:42:05.612638 5158 gossip/gossip.go:1055  [n?] no resolvers found; use --join to specify a connected node
I161106 07:42:05.612873 5158 base/node_id.go:62  NodeID set to 1
I161106 07:42:05.612944 5158 storage/engine/rocksdb.go:340  opening in memory rocksdb instance
I161106 07:42:05.625734 5158 storage/store.go:1174  [n1,s1]: failed initial metrics computation: [n1,s1]: system config not yet available
I161106 07:42:05.628101 5162 storage/replica_proposal.go:334  [s1,r1/1:/M{in-ax}] new range lease replica {1 1 1} 1970-01-01 00:00:00 +0000 UTC 900.000124ms following replica {0 0 0} 1970-01-01 00:00:00 +0000 UTC 0s [physicalTime=1970-01-01 00:00:00.000000123 +0000 UTC]
I161106 07:42:05.631643 5158 util/stop/stopper.go:396  stop has been called, stopping or quiescing all running tasks
	replica_test.go:5655: expected an ambiguous result error; got <nil>
```

This test fails absolutely instantly under stress; I'm going to skip it.

```
make stress STRESSFLAGS='-maxfails 1 -stderr -p 16' PKG=./pkg/storage TESTS=TestReplicaCancelRaft
go list -tags '' -f \
	'go test -v  -tags '\'''\'' -ldflags '\'''\'' -i -c {{.ImportPath}} -o {{.Dir}}/stress.test && (cd {{.Dir}} && if [ -f stress.test ]; then stress -maxfails 1 -stderr -p 16 ./stress.test -test.run '\''TestReplicaCancelRaft'\'' -test.timeout 2m ; fi)' ./pkg/storage | \
	/usr/local/bin/bash
github.com/cockroachdb/cockroach/pkg/storage
github.com/cockroachdb/cockroach/pkg/storage_test
testmain

I161110 15:20:45.643373 1 rand.go:76  Random seed: 637254767943773797
I161110 15:20:45.646010 20 gossip/gossip.go:237  [n?] initial resolvers: []
W161110 15:20:45.646049 20 gossip/gossip.go:1055  [n?] no resolvers found; use --join to specify a connected node
I161110 15:20:45.646103 20 base/node_id.go:62  NodeID set to 1
I161110 15:20:45.646173 20 storage/engine/rocksdb.go:340  opening in memory rocksdb instance
I161110 15:20:45.668572 14 storage/replica_proposal.go:328  [s1,r1/1:/M{in-ax}] new range lease replica {1 1 1} 1970-01-01 00:00:00 +0000 UTC 900.000124ms following replica {0 0 0} 1970-01-01 00:00:00 +0000 UTC 0s [physicalTime=1970-01-01 00:00:00.000000123 +0000 UTC]
I161110 15:20:45.672998 20 util/stop/stopper.go:396  stop has been called, stopping or quiescing all running tasks
--- FAIL: TestReplicaCancelRaft (0.03s)
	replica_test.go:5776: expected an ambiguous result error; got <nil>
I161110 15:20:45.674907 84 gossip/gossip.go:237  [n?] initial resolvers: []
W161110 15:20:45.674932 84 gossip/gossip.go:1055  [n?] no resolvers found; use --join to specify a connected node
I161110 15:20:45.674969 84 base/node_id.go:62  NodeID set to 1
I161110 15:20:45.674992 84 storage/engine/rocksdb.go:340  opening in memory rocksdb instance
I161110 15:20:45.676812 84 storage/store.go:1187  [n1,s1]: failed initial metrics computation: [n1,s1]: system config not yet available
I161110 15:20:45.677602 94 storage/replica_proposal.go:328  [s1,r1/1:/M{in-ax}] new range lease replica {1 1 1} 1970-01-01 00:00:00 +0000 UTC 900.000124ms following replica {0 0 0} 1970-01-01 00:00:00 +0000 UTC 0s [physicalTime=1970-01-01 00:00:00.000000123 +0000 UTC]
I161110 15:20:45.678034 84 storage/replica_test.go:6137  abandoning command 0
I161110 15:20:45.678082 84 storage/replica_test.go:6137  abandoning command 2
I161110 15:20:45.678096 84 storage/replica_test.go:6137  abandoning command 3
I161110 15:20:45.678137 84 storage/replica_test.go:6137  abandoning command 6
I161110 15:20:45.678197 84 storage/replica_test.go:6137  abandoning command 9
I161110 15:20:45.678792 84 util/stop/stopper.go:396  stop has been called, stopping or quiescing all running tasks
FAIL


ERROR: exit status 1

7 runs completed, 1 failures, over 0s
FAIL
make: *** [stress] Error 1
```

--- a/pkg/storage/replica.go
+++ b/pkg/storage/replica.go
@@ -1220,6 +1220,60 @@ func (r *Replica) checkBatchRequest(ba roachpb.BatchRequest) error {
 	return nil
 }
 
+// endCmds holds necessary information to end a batch after Raft
+// command processing.
+type endCmds struct {
+	repl      *Replica
+	cmdGlobal *cmd
+	cmdLocal  *cmd
+	ba        roachpb.BatchRequest
+}
+
+// done removes pending commands from the command queue and updates
+// the timestamp cache using the final timestamp of each command.  The
+// returned error replaces the supplied error.
+func (ec *endCmds) done(br *roachpb.BatchResponse, pErr *roachpb.Error, shouldRetry bool) {
+	// Update the timestamp cache if the command succeeded and is not
+	// being retried. Each request is considered in turn; only those
+	// marked as affecting the cache are processed. Inconsistent reads
+	// are excluded.
+	if pErr == nil && !shouldRetry && ec.ba.ReadConsistency != roachpb.INCONSISTENT {
+		cr := cacheRequest{
+			timestamp: ec.ba.Timestamp,
+			txnID:     ec.ba.GetTxnID(),
+		}
+
+		for _, union := range ec.ba.Requests {
+			args := union.GetInner()
+			if updatesTimestampCache(args) {
+				header := args.Header()
+				switch args.(type) {
+				case *roachpb.DeleteRangeRequest:
+					// DeleteRange adds to the write timestamp cache to prevent
+					// subsequent writes from rewriting history.
+					cr.writes = append(cr.writes, header)
+				case *roachpb.EndTransactionRequest:
+					// EndTransaction adds to the write timestamp cache to ensure replays
+					// create a transaction record with WriteTooOld set.
+					key := keys.TransactionKey(header.Key, *cr.txnID)
+					cr.txn = roachpb.Span{Key: key}
+				default:
+					cr.reads = append(cr.reads, header)
+				}
+			}
+		}
+
+		ec.repl.mu.Lock()
+		ec.repl.mu.tsCache.AddRequest(cr)
+		ec.repl.mu.Unlock()
+	}
+
+	ec.repl.cmdQMu.Lock()
+	ec.repl.cmdQMu.global.remove(ec.cmdGlobal)
+	ec.repl.cmdQMu.local.remove(ec.cmdLocal)
+	ec.repl.cmdQMu.Unlock()
+}
+
 // beginCmds waits for any overlapping, already-executing commands via
 // the command queue and adds itself to queues based on keys affected by the
 // batched commands. This gates subsequent commands with overlapping keys or
@@ -1227,9 +1281,7 @@ func (r *Replica) checkBatchRequest(ba roachpb.BatchRequest) error {
 // already in the queue. Returns a cleanup function to be called when the
 // commands are done and can be removed from the queue, and whose returned
 // error is to be used in place of the supplied error.
-func (r *Replica) beginCmds(
-	ctx context.Context, ba *roachpb.BatchRequest,
-) (func(*roachpb.BatchResponse, *roachpb.Error, bool) *roachpb.Error, error) {
+func (r *Replica) beginCmds(ctx context.Context, ba *roachpb.BatchRequest) (*endCmds, error) {
 	var cmdGlobal *cmd
 	var cmdLocal *cmd
 	// Don't use the command queue for inconsistent reads.
@@ -1348,62 +1400,13 @@ func (r *Replica) beginCmds(
 		}
 	}
 
-	return func(br *roachpb.BatchResponse, pErr *roachpb.Error, shouldRetry bool) *roachpb.Error {
-		return r.endCmds(cmdGlobal, cmdLocal, ba, br, pErr, shouldRetry)
-	}, nil
-}
-
-// endCmds removes pending commands from the command queue and updates
-// the timestamp cache using the final timestamp of each command.
-// The returned error replaces the supplied error.
-func (r *Replica) endCmds(
-	cmdGlobal *cmd,
-	cmdLocal *cmd,
-	ba *roachpb.BatchRequest,
-	br *roachpb.BatchResponse,
-	pErr *roachpb.Error,
-	shouldRetry bool,
-) *roachpb.Error {
-	// Update the timestamp cache if the command succeeded and is not
-	// being retried. Each request is considered in turn; only those
-	// marked as affecting the cache are processed. Inconsistent reads
-	// are excluded.
-	if pErr == nil && !shouldRetry && ba.ReadConsistency != roachpb.INCONSISTENT {
-		cr := cacheRequest{
-			timestamp: ba.Timestamp,
-			txnID:     ba.GetTxnID(),
-		}
-
-		for _, union := range ba.Requests {
-			args := union.GetInner()
-			if updatesTimestampCache(args) {
-				header := args.Header()
-				switch args.(type) {
-				case *roachpb.DeleteRangeRequest:
-					// DeleteRange adds to the write timestamp cache to prevent
-					// subsequent writes from rewriting history.
-					cr.writes = append(cr.writes, header)
-				case *roachpb.EndTransactionRequest:
-					// EndTransaction adds to the write timestamp cache to ensure replays
-					// create a transaction record with WriteTooOld set.
-					key := keys.TransactionKey(header.Key, *cr.txnID)
-					cr.txn = roachpb.Span{Key: key}
-				default:
-					cr.reads = append(cr.reads, header)
-				}
-			}
-		}
-
-		r.mu.Lock()
-		r.mu.tsCache.AddRequest(cr)
-		r.mu.Unlock()
+	ec := &endCmds{
+		repl:      r,
+		cmdGlobal: cmdGlobal,
+		cmdLocal:  cmdLocal,
+		ba:        *ba,
 	}
-
-	r.cmdQMu.Lock()
-	r.cmdQMu.global.remove(cmdGlobal)
-	r.cmdQMu.local.remove(cmdLocal)
-	r.cmdQMu.Unlock()
-	return pErr
+	return ec, nil
 }
 
 // applyTimestampCache moves the batch timestamp forward depending on
@@ -1574,16 +1577,14 @@ func (r *Replica) addReadOnlyCmd(
 		}
 	}
 
-	endCmdsFunc := func(_ *roachpb.BatchResponse, pErr *roachpb.Error, _ bool) *roachpb.Error {
-		return pErr
-	}
+	var endCmds *endCmds
 
 	if !ba.IsNonKV() {
 		// Add the read to the command queue to gate subsequent
 		// overlapping commands until this command completes.
 		log.Event(ctx, "command queue")
 		var err error
-		endCmdsFunc, err = r.beginCmds(ctx, &ba)
+		endCmds, err = r.beginCmds(ctx, &ba)
 		if err != nil {
 			return nil, roachpb.NewError(err)
 		}
@@ -1598,7 +1599,9 @@ func (r *Replica) addReadOnlyCmd(
 	// timestamp cache update is synchronized. This is wrapped to delay
 	// pErr evaluation to its value when returning.
 	defer func() {
-		pErr = endCmdsFunc(br, pErr, false)
+		if endCmds != nil {
+			endCmds.done(br, pErr, false)
+		}
 	}()
 
 	r.mu.Lock()
@@ -1706,6 +1709,7 @@ func (r *Replica) tryAddWriteCmd(
 	ctx context.Context, ba roachpb.BatchRequest,
 ) (br *roachpb.BatchResponse, pErr *roachpb.Error, shouldRetry bool) {
 	isNonKV := ba.IsNonKV()
+	var endCmds *endCmds
 	if !isNonKV {
 		// Add the write to the command queue to gate subsequent overlapping
 		// commands until this command completes. Note that this must be
@@ -1713,7 +1717,8 @@ func (r *Replica) tryAddWriteCmd(
 		// timestamp cache is only updated after preceding commands have
 		// been run to successful completion.
 		log.Event(ctx, "command queue")
-		endCmdsFunc, err := r.beginCmds(ctx, &ba)
+		var err error
+		endCmds, err = r.beginCmds(ctx, &ba)
 		if err != nil {
 			return nil, roachpb.NewError(err), false
 		}
@@ -1721,7 +1726,9 @@ func (r *Replica) tryAddWriteCmd(
 		// Guarantee we remove the commands from the command queue. This is
 		// wrapped to delay pErr evaluation to its value when returning.
 		defer func() {
-			pErr = endCmdsFunc(br, pErr, shouldRetry)
+			if endCmds != nil {
+				endCmds.done(br, pErr, shouldRetry)
+			}
 		}()
 	}
 
@@ -1779,7 +1786,7 @@ func (r *Replica) tryAddWriteCmd(
 
 	log.Event(ctx, "raft")
 
-	ch, tryAbandon, err := r.propose(ctx, ba)
+	ch, tryAbandon, err := r.propose(ctx, ba, endCmds)
 	if err != nil {
 		return nil, roachpb.NewError(err), false
 	}
@@ -1789,23 +1796,31 @@ func (r *Replica) tryAddWriteCmd(
 	shouldQuiesce := r.store.stopper.ShouldQuiesce()
 	for {
 		select {
-		case respWithErr := <-ch:
-			return respWithErr.Reply, respWithErr.Err, respWithErr.ShouldRetry
+		case propResult := <-ch:
+			endCmds = nil // these will have been invoked post-Raft.
+			return propResult.Reply, propResult.Err, propResult.ShouldRetry
 		case <-ctxDone:
-			// We ignore cancelled contexts for running commands in order to
-			// properly update the timestamp and command queue. Exiting early,
-			// before knowing the final disposition of the command would make
-			// those updates impossible, leading to potential inconsistencies.
-			// TODO(spencer): move updates to the timestamp cache to raft
-			// command execution.
-			log.Warningf(ctx, "ignoring cancelled context for command %s", ba)
+			// If our context was cancelled, return an AmbiguousResultError
+			// to indicate to caller that the command may have executed.
+			// However, we proceed only if the command isn't already being
+			// executed and using our context, in which case we expect it to
+			// finish soon.
+			if tryAbandon() {
+				log.Warningf(ctx, "context cancellation of command %s", ba)
+				endCmds = nil
+				return nil, roachpb.NewError(roachpb.NewAmbiguousResultError()), false
+			}
 			ctxDone = nil
 		case <-shouldQuiesce:
 			// If we're shutting down, return an AmbiguousResultError to
 			// indicate to caller that the command may have executed.
-			// However, we proceed only if the command isn't already being
-			// executed and using our context, in which case we expect it to
-			// finish soon.
+			// tryAbandon indicates whether the command is already executing
+			// with our context. Note that in the shutdown case, we *do not*
+			// set the endCmds var to nil. We have no expectation during
+			// shutdown that Raft will continue processing, so we need to
+			// free up the command queue so other waiters can proceed. This
+			// is safe because with the stopper quiescing, no new commands
+			// can be proposed to Raft successfully.
 			if tryAbandon() {
 				log.Warningf(ctx, "shutdown cancellation of command %s", ba)
 				return nil, roachpb.NewError(roachpb.NewAmbiguousResultError()), false
@@ -1822,16 +1837,18 @@ func (r *Replica) requestToProposal(
 	idKey storagebase.CmdIDKey,
 	replica roachpb.ReplicaDescriptor,
 	ba roachpb.BatchRequest,
+	endCmds *endCmds,
 ) (*ProposalData, *roachpb.Error) {
 	if propEvalKV {
-		return r.evaluateProposal(ctx, idKey, replica, ba)
+		return r.evaluateProposal(ctx, idKey, replica, ba, endCmds)
 	}
 	return &ProposalData{
 		Cmd: &ba,
 		LocalProposalData: LocalProposalData{
-			ctx:   ctx,
-			idKey: idKey,
-			done:  make(chan proposalResult, 1),
+			ctx:     ctx,
+			idKey:   idKey,
+			endCmds: endCmds,
+			doneCh:  make(chan proposalResult, 1),
 		},
 	}, nil
 }
@@ -1851,6 +1868,7 @@ func (r *Replica) evaluateProposal(
 	idKey storagebase.CmdIDKey,
 	replica roachpb.ReplicaDescriptor,
 	ba roachpb.BatchRequest,
+	endCmds *endCmds,
 ) (*ProposalData, *roachpb.Error) {
 	// Note that we don't hold any locks at this point. This is important
 	// since evaluating a proposal is expensive (at least under proposer-
@@ -1884,7 +1902,8 @@ func (r *Replica) evaluateProposal(
 
 	pd.ctx = ctx
 	pd.idKey = idKey
-	pd.done = make(chan proposalResult, 1)
+	pd.endCmds = endCmds
+	pd.doneCh = make(chan proposalResult, 1)
 	pd.IsLeaseRequest = ba.IsLeaseRequest()
 	pd.IsFreeze = ba.IsFreeze()
 	pd.IsConsistencyRelated = ba.IsConsistencyRelated()
@@ -1945,7 +1964,7 @@ func makeIDKey() storagebase.CmdIDKey {
 // - any error obtained during the creation or proposal of the command, in
 //   which case the other returned values are zero.
 func (r *Replica) propose(
-	ctx context.Context, ba roachpb.BatchRequest,
+	ctx context.Context, ba roachpb.BatchRequest, endCmds *endCmds,
 ) (chan proposalResult, func() bool, error) {
 	r.mu.Lock()
 	if err := r.mu.destroyed; err != nil {
@@ -1976,13 +1995,16 @@ func (r *Replica) propose(
 	defer r.raftMu.Unlock()
 
 	idKey := makeIDKey()
-	pCmd, pErr := r.requestToProposal(ctx, idKey, repDesc, ba)
+	pCmd, pErr := r.requestToProposal(ctx, idKey, repDesc, ba, endCmds)
 	// An error here corresponds to a failfast-proposal: The command resulted
 	// in an error and did not need to commit a batch (the common error case).
 	if pErr != nil {
 		r.handleProposalData(
 			ctx, repDesc, pCmd.LocalProposalData, pCmd.ReplicatedProposalData,
 		)
+		if endCmds != nil {
+			endCmds.done(nil, pErr, false)
+		}
 		ch := make(chan proposalResult, 1)
 		ch <- proposalResult{Err: pErr}
 		close(ch)
@@ -2006,7 +2028,7 @@ func (r *Replica) propose(
 		r.mu.Unlock()
 		return ok
 	}
-	return pCmd.done, tryAbandon, nil
+	return pCmd.doneCh, tryAbandon, nil
 }
 
 // submitProposalLocked proposes or re-proposes a command in r.mu.proposals.
@@ -2696,8 +2718,7 @@ func (r *Replica) refreshProposalsLocked(refreshAtDelta int, reason refreshRaftR
 		// The command's designated lease index range was filled up, so send it
 		// back to the proposer for a retry.
 		log.Eventf(p.ctx, "retry proposal %x: %s", p.idKey, reason)
-		p.done <- proposalResult{ShouldRetry: true}
-		close(p.done)
+		p.finish(proposalResult{ShouldRetry: true})
 		numShouldRetry++
 	}
 	if log.V(1) && (numShouldRetry > 0 || len(reproposals) > 0) {
@@ -2720,8 +2741,7 @@ func (r *Replica) refreshProposalsLocked(refreshAtDelta int, reason refreshRaftR
 		log.Eventf(p.ctx, "re-submitting command %x to Raft: %s", p.idKey, reason)
 		if err := r.submitProposalLocked(p); err != nil {
 			delete(r.mu.proposals, p.idKey)
-			p.done <- proposalResult{Err: roachpb.NewError(err)}
-			close(p.done)
+			p.finish(proposalResult{Err: roachpb.NewError(err)})
 		}
 	}
 }
@@ -3053,15 +3073,17 @@ func (r *Replica) processRaftCommand(
 			// use the context any more once we signal the client, so we make
 			// sure we signal it at the end of this method, when the context
 			// has been fully used.
-			defer func(ch chan proposalResult) {
+			copyCmd := *cmd
+			// Clear the endCmds and doneCh so that ProposalData.finish()
+			// is a noop when invoked below.
+			cmd.endCmds = nil
+			cmd.doneCh = make(chan proposalResult, 1)
+			defer func() {
 				// Assert against another defer trying to use the context after
 				// the client has been signaled.
 				ctx = nil
-
-				ch <- proposalResult{ShouldRetry: true}
-				close(ch)
-			}(cmd.done)
-			cmd.done = make(chan proposalResult, 1)
+				copyCmd.finish(proposalResult{ShouldRetry: true})
+			}()
 		}
 	}
 	// When frozen, the Range only applies freeze- and consistency-related
@@ -3102,6 +3124,7 @@ func (r *Replica) processRaftCommand(
 				idKey,
 				raftCmd.OriginReplica,
 				*raftCmd.Cmd,
+				nil,
 			)
 			// Then, change the raftCmd to reflect the result of the
 			// evaluation, filling in the ProposalData (which is now properly
@@ -3114,9 +3137,11 @@ func (r *Replica) processRaftCommand(
 			raftCmd.ReplicatedProposalData = &innerPD.ReplicatedProposalData
 			writeBatch = innerPD.WriteBatch
 			if cmdProposedLocally {
-				done := cmd.LocalProposalData.done
+				endCmds := cmd.LocalProposalData.endCmds
+				doneCh := cmd.LocalProposalData.doneCh
 				cmd.LocalProposalData = innerPD.LocalProposalData
-				cmd.done = done
+				cmd.endCmds = endCmds
+				cmd.doneCh = doneCh
 				cmd.ctx = nil // already have ctx
 			}
 			// Proposals which would failfast with proposer-evaluated KV now
@@ -3191,8 +3216,7 @@ func (r *Replica) processRaftCommand(
 	}
 
 	if cmdProposedLocally {
-		cmd.done <- response
-		close(cmd.done)
+		cmd.finish(response)
 	} else if response.Err != nil {
 		log.VEventf(ctx, 1, "error executing raft command %s: %s", raftCmd.Cmd, response.Err)
 	}

--- a/pkg/storage/replica_proposal.go
+++ b/pkg/storage/replica_proposal.go
@@ -57,7 +57,10 @@ type LocalProposalData struct {
 	// the common case, this field is nil.
 	Err   *roachpb.Error
 	Reply *roachpb.BatchResponse
-	done  chan proposalResult // Used to signal waiting RPC handler
+	// Called after command execution to update the timestamp cache &
+	// command queue.
+	endCmds *endCmds
+	doneCh  chan proposalResult // Used to signal waiting RPC handler
 
 	Batch engine.Batch
 
@@ -91,6 +94,20 @@ type LocalProposalData struct {
 	maybeAddToSplitQueue bool
 	// Call maybeGossipNodeLiveness with the specified Span, if set.
 	maybeGossipNodeLiveness *roachpb.Span
+}
+
+// finish first invokes the endCmds function and then sends the
+// specified proposalResult on the lpd's done channel. endCmds is
+// invoked here in order to allow the original client to be cancelled
+// and possibly no longer listening to this done channel, and so can't
+// be counted on to invoke endCmds itself.
+func (lpd *LocalProposalData) finish(pr proposalResult) {
+	if lpd.endCmds != nil {
+		lpd.endCmds.done(pr.Reply, pr.Err, pr.ShouldRetry)
+		lpd.endCmds = nil
+	}
+	lpd.doneCh <- pr
+	close(lpd.doneCh)
 }
 
 // ProposalData is the result of preparing a Raft proposal. That is, the
@@ -578,7 +595,8 @@ func (r *Replica) handleLocalProposalData(
 	{
 		lpd.idKey = storagebase.CmdIDKey("")
 		lpd.Batch = nil
-		lpd.done = nil
+		lpd.endCmds = nil
+		lpd.doneCh = nil
 		lpd.ctx = nil
 		lpd.Err = nil
 		lpd.proposedAtTicks = 0

--- a/pkg/storage/replica_test.go
+++ b/pkg/storage/replica_test.go
@@ -430,7 +430,7 @@ func sendLeaseRequest(r *Replica, l *roachpb.Lease) error {
 	ba := roachpb.BatchRequest{}
 	ba.Timestamp = r.store.Clock().Now()
 	ba.Add(&roachpb.RequestLeaseRequest{Lease: *l})
-	ch, _, err := r.propose(context.TODO(), ba)
+	ch, _, err := r.propose(context.TODO(), ba, nil)
 	if err == nil {
 		// Next if the command was committed, wait for the range to apply it.
 		// TODO(bdarnell): refactor this to a more conventional error-handling pattern.
@@ -1099,7 +1099,7 @@ func TestReplicaLeaseRejectUnknownRaftNodeID(t *testing.T) {
 	ba := roachpb.BatchRequest{}
 	ba.Timestamp = tc.repl.store.Clock().Now()
 	ba.Add(&roachpb.RequestLeaseRequest{Lease: *lease})
-	ch, _, err := tc.repl.propose(context.Background(), ba)
+	ch, _, err := tc.repl.propose(context.Background(), ba, nil)
 	if err == nil {
 		// Next if the command was committed, wait for the range to apply it.
 		// TODO(bdarnell): refactor to a more conventional error-handling pattern.
@@ -1781,9 +1781,7 @@ func TestLeaseConcurrent(t *testing.T) {
 						// otherwise its context (and tracing span) may be used after the
 						// client cleaned up.
 						delete(tc.repl.mu.proposals, cmd.idKey)
-						cmd.done <- proposalResult{
-							Err: roachpb.NewErrorf(origMsg),
-						}
+						cmd.finish(proposalResult{Err: roachpb.NewErrorf(origMsg)})
 						return
 					}
 					if err := defaultSubmitProposalLocked(tc.repl, cmd); err != nil {
@@ -3263,7 +3261,7 @@ func TestRaftReplayProtectionInTxn(t *testing.T) {
 		// Reach in and manually send to raft (to simulate Raft replay) and
 		// also avoid updating the timestamp cache; verify WriteTooOldError.
 		ba.Timestamp = txn.OrigTimestamp
-		ch, _, err := tc.repl.propose(context.Background(), ba)
+		ch, _, err := tc.repl.propose(context.Background(), ba, nil)
 		if err != nil {
 			t.Fatalf("%d: unexpected error: %s", i, err)
 		}
@@ -5753,28 +5751,66 @@ func TestGCIncorrectRange(t *testing.T) {
 
 // TestReplicaCancelRaft checks that it is possible to safely abandon Raft
 // commands via a cancelable context.Context.
+
 func TestReplicaCancelRaft(t *testing.T) {
 	defer leaktest.AfterTest(t)()
-	t.Skip("TODO(spencerkimball): https://github.com/cockroachdb/cockroach/issues/10488")
-	// Pick a key unlikely to be used by background processes.
-	key := []byte("acdfg")
-	tc := testContext{}
-	tc.Start(t)
-	defer tc.Stop()
-	var ba roachpb.BatchRequest
-	ba.Add(&roachpb.GetRequest{
-		Span: roachpb.Span{Key: key},
-	})
-	if err := ba.SetActiveTimestamp(tc.Clock().Now); err != nil {
-		t.Fatal(err)
-	}
-	// Quiesce the stopper. Note that calling `Stop()` here would let the
-	// test panic as we still need the engine open (at least with proposer-
-	// evaluated KV).
-	tc.stopper.Quiesce()
-	_, pErr := tc.repl.addWriteCmd(context.Background(), ba)
-	if _, ok := pErr.GetDetail().(*roachpb.AmbiguousResultError); !ok {
-		t.Fatalf("expected an ambiguous result error; got %v", pErr)
+	for _, cancelEarly := range []bool{true, false} {
+		func() {
+			// Pick a key unlikely to be used by background processes.
+			key := []byte("acdfg")
+			ctx, cancel := context.WithCancel(context.Background())
+			cfg := TestStoreConfig(nil)
+			if !cancelEarly {
+				cfg.TestingKnobs.TestingCommandFilter =
+					func(filterArgs storagebase.FilterArgs) *roachpb.Error {
+						if !filterArgs.Req.Header().Key.Equal(key) {
+							return nil
+						}
+						cancel()
+						return nil
+					}
+
+			}
+			tc := testContext{}
+			tc.StartWithStoreConfig(t, cfg)
+			defer tc.Stop()
+			if cancelEarly {
+				cancel()
+				tc.repl.mu.Lock()
+				tc.repl.mu.submitProposalFn = func(*ProposalData) error {
+					return nil
+				}
+				tc.repl.mu.Unlock()
+			}
+			var ba roachpb.BatchRequest
+			ba.Add(&roachpb.GetRequest{
+				Span: roachpb.Span{Key: key},
+			})
+			if err := ba.SetActiveTimestamp(tc.Clock().Now); err != nil {
+				t.Fatal(err)
+			}
+			br, pErr := tc.repl.addWriteCmd(ctx, ba)
+			if pErr == nil {
+				if !cancelEarly {
+					// We cancelled the context while the command was already
+					// being processed, so the client had to wait for successful
+					// execution.
+					return
+				}
+				t.Fatalf("expected an error, but got successful response %+v", br)
+			}
+			// If we cancelled the context early enough, we expect to receive a
+			// corresponding error and not wait for the command.
+			if cancelEarly {
+				if !testutils.IsPError(pErr, context.Canceled.Error()) {
+					t.Errorf("unexpected error: %s", pErr)
+				}
+			} else {
+				if _, ok := pErr.GetDetail().(*roachpb.AmbiguousResultError); !ok {
+					t.Errorf("expected an ambiguous result error; got %v", pErr)
+				}
+			}
+		}()
 	}
 }
 
@@ -5990,7 +6026,7 @@ func TestReplicaIDChangePending(t *testing.T) {
 			Key: roachpb.Key("a"),
 		},
 	})
-	_, _, err := repl.propose(context.Background(), ba)
+	_, _, err := repl.propose(context.Background(), ba, nil)
 	if err != nil {
 		t.Fatal(err)
 	}
@@ -6123,7 +6159,7 @@ func TestReplicaCancelRaftCommandProgress(t *testing.T) {
 			ba.Add(&roachpb.PutRequest{Span: roachpb.Span{
 				Key: roachpb.Key(fmt.Sprintf("k%d", i))}})
 			cmd, pErr := repl.requestToProposal(
-				context.Background(), makeIDKey(), repDesc, ba,
+				context.Background(), makeIDKey(), repDesc, ba, nil,
 			)
 			if pErr != nil {
 				t.Fatal(pErr)
@@ -6140,7 +6176,7 @@ func TestReplicaCancelRaftCommandProgress(t *testing.T) {
 			} else if err := repl.submitProposalLocked(cmd); err != nil {
 				t.Error(err)
 			} else {
-				chs = append(chs, cmd.done)
+				chs = append(chs, cmd.doneCh)
 			}
 			repl.mu.Unlock()
 		}
@@ -6198,7 +6234,7 @@ func TestReplicaBurstPendingCommandsAndRepropose(t *testing.T) {
 			ba.Timestamp = tc.Clock().Now()
 			ba.Add(&roachpb.PutRequest{Span: roachpb.Span{
 				Key: roachpb.Key(fmt.Sprintf("k%d", i))}})
-			cmd, pErr := tc.repl.requestToProposal(ctx, makeIDKey(), repDesc, ba)
+			cmd, pErr := tc.repl.requestToProposal(ctx, makeIDKey(), repDesc, ba, nil)
 			if pErr != nil {
 				t.Fatal(pErr)
 			}
@@ -6206,7 +6242,7 @@ func TestReplicaBurstPendingCommandsAndRepropose(t *testing.T) {
 			tc.repl.raftMu.Lock()
 			tc.repl.mu.Lock()
 			tc.repl.insertProposalLocked(cmd, repDesc)
-			chs = append(chs, cmd.done)
+			chs = append(chs, cmd.doneCh)
 			tc.repl.mu.Unlock()
 			tc.repl.raftMu.Unlock()
 		}
@@ -6315,8 +6351,9 @@ func TestReplicaRefreshPendingCommandsTicks(t *testing.T) {
 		var ba roachpb.BatchRequest
 		ba.Timestamp = tc.Clock().Now()
 		ba.Add(&roachpb.PutRequest{Span: roachpb.Span{Key: roachpb.Key(id)}})
-		cmd, pErr := r.requestToProposal(context.Background(),
-			storagebase.CmdIDKey(id), repDesc, ba)
+		cmd, pErr := r.requestToProposal(
+			context.Background(), storagebase.CmdIDKey(id), repDesc, ba, nil,
+		)
 		if pErr != nil {
 			t.Fatal(pErr)
 		}

--- a/pkg/storage/store.go
+++ b/pkg/storage/store.go
@@ -2009,7 +2009,7 @@ func (s *Store) removeReplicaImpl(
 			Err:   roachpb.NewError(roachpb.NewRangeNotFoundError(rep.RangeID)),
 		}
 		for _, p := range rep.mu.proposals {
-			p.done <- resp
+			p.finish(resp)
 		}
 	}
 	// Clear the map.

--- a/pkg/storage/store_test.go
+++ b/pkg/storage/store_test.go
@@ -513,7 +513,7 @@ func TestStoreRemoveReplicaDestroy(t *testing.T) {
 	}
 
 	if _, _, err := repl1.propose(
-		context.Background(), roachpb.BatchRequest{},
+		context.Background(), roachpb.BatchRequest{}, nil,
 	); err != expErr {
 		t.Fatalf("expected error %s, but got %v", expErr, err)
 	}

